File: windows/internal/platform/toastxml.go
package platform

import (
	"fmt"
	"strings"
)

// toastXML is a notification as Windows' toast XML schema
// (learn.microsoft.com/windows/apps/design/shell/tiles-and-notifications/toast-schema).
// Clicks and buttons use protocol activation (https: links, or droplet:
// links handled by a new droplet.exe), so nothing has to stay registered as
// a COM server to receive them.
func toastXML(n Notification, actionKey string) string {
	var x strings.Builder
	x.WriteString(`<toast`)
	if n.Click != nil {
		fmt.Fprintf(&x, ` activationType="protocol" launch="%s"`, xmlEsc(ActionURL(actionKey, *n.Click)))
	}
	if n.Urgent {
		x.WriteString(` scenario="incomingCall"`)
	}
	x.WriteString(`><visual><binding template="ToastGeneric">`)
	fmt.Fprintf(&x, `<text hint-maxLines="1">%s</text>`, xmlEsc(clip(n.Title, 120)))
	if n.Body != "" {
		fmt.Fprintf(&x, `<text>%s</text>`, xmlEsc(clip(n.Body, 400)))
	}
	x.WriteString(`</binding></visual>`)
	if len(n.Buttons) > 0 {
		x.WriteString(`<actions>`)
		for _, b := range n.Buttons {
			fmt.Fprintf(&x, `<action content="%s" activationType="protocol" arguments="%s"/>`,
				xmlEsc(b.Label), xmlEsc(ActionURL(actionKey, b)))
		}
		x.WriteString(`</actions>`)
	}
	// rings have their own (louder, looping) sound; the rest keep the default chime
	if n.Urgent {
		x.WriteString(`<audio silent="true"/>`)
	}
	x.WriteString(`</toast>`)
	return x.String()
}

// toastTag is a notification's tag as Windows accepts it: at most 64
// UTF-16 units (60 here, to be safe), cut between characters.
func toastTag(t string) string {
	units := 0
	for i, r := range t {
		n := 1
		if r > 0xFFFF {
			n = 2 // a surrogate pair
		}
		if units+n > 60 {
			return t[:i]
		}
		units += n
	}
	return t
}

func clip(s string, n int) string {
	r := []rune(s)
	if len(r) > n {
		return string(r[:n-1]) + "…"
	}
	return s
}

// xmlEsc makes s safe as XML text or an attribute value. Characters XML
// can't carry at all (most control characters) are dropped, since one of
// them in a message would make Windows reject the whole notification.
func xmlEsc(s string) string {
	var b strings.Builder
	b.Grow(len(s))
	for _, r := range s { // invalid UTF-8 arrives here as U+FFFD
		switch {
		case r == '&':
			b.WriteString("&amp;")
		case r == '<':
			b.WriteString("&lt;")
		case r == '>':
			b.WriteString("&gt;")
		case r == '"':
			b.WriteString("&quot;")
		case r == '\'':
			b.WriteString("&apos;")
		case r == '\t', r == '\n', r == '\r':
			b.WriteRune(r)
		case r < 0x20, r == 0xFFFE, r == 0xFFFF:
			// not allowed in XML 1.0
		default:
			b.WriteRune(r)
		}
	}
	return b.String()
}

File: windows-net/src/Droplet.Windows/Platform/ToastXml.cs
using System.Globalization;
using System.Text;
using Droplet.Core.Platform;
using Droplet.Windows.Shell;

namespace Droplet.Windows.Platform;

/// <summary>
/// A notification as Windows' toast XML (learn.microsoft.com/windows/apps/design/shell/
/// tiles-and-notifications/toast-schema). Clicks and buttons use protocol activation:
/// https links, or <c>droplet:</c> links (<see cref="ActionLinks"/>). Pure, so it's tested
/// anywhere; a port of the Go app's <c>toastxml.go</c>.
/// </summary>
internal static class ToastXml
{
    /// <summary>The XML for <paramref name="n"/>.</summary>
    /// <param name="n">The notification.</param>
    /// <param name="key">The action key for droplet: links.</param>
    /// <param name="appLogo">A file path for the app logo, or null.</param>
    public static string Build(Notification n, string key, string? appLogo = null)
    {
        ArgumentNullException.ThrowIfNull(n);
        var x = new StringBuilder("<toast");
        if (n.Click is { } click)
        {
            x.Append(CultureInfo.InvariantCulture, $" activationType=\"protocol\" launch=\"{Esc(ActionLinks.ToUrl(click, key))}\"");
        }
        if (n.Urgent)
        {
            // stays up until answered; rings have their own, louder, looping sound
            x.Append(" scenario=\"incomingCall\"");
        }
        x.Append("><visual><binding template=\"ToastGeneric\">");
        x.Append(CultureInfo.InvariantCulture, $"<text hint-maxLines=\"1\">{Esc(Clip(n.Title, 120))}</text>");
        if (n.Body.Length > 0)
        {
            x.Append(CultureInfo.InvariantCulture, $"<text>{Esc(Clip(n.Body, 400))}</text>");
        }
        if (!string.IsNullOrEmpty(n.App))
        {
            x.Append(CultureInfo.InvariantCulture, $"<text placement=\"attribution\">{Esc(Clip(n.App, 60))}</text>");
        }
        if (!string.IsNullOrEmpty(appLogo))
        {
            x.Append(CultureInfo.InvariantCulture, $"<image placement=\"appLogoOverride\" hint-crop=\"none\" src=\"{Esc(new Uri(appLogo).AbsoluteUri)}\"/>");
        }
        x.Append("</binding></visual>");
        // one that stays up needs a way to dismiss it
        var buttons = n.Buttons.Count > 0 || !n.Urgent ? n.Buttons : [new NotificationAction("Dismiss", NotificationActionKind.OpenUrl)];
        if (buttons.Count > 0)
        {
            x.Append("<actions>");
            foreach (var b in buttons.Take(5))
            {
                if (b.Kind == NotificationActionKind.OpenUrl && b.Arg.Length == 0)
                {
                    x.Append(CultureInfo.InvariantCulture, $"<action content=\"{Esc(b.Label)}\" arguments=\"dismiss\" activationType=\"system\"/>");
                    continue;
                }
                x.Append(CultureInfo.InvariantCulture, $"<action content=\"{Esc(b.Label)}\" activationType=\"protocol\" arguments=\"{Esc(ActionLinks.ToUrl(b, key))}\"/>");
            }
            x.Append("</actions>");
        }
        if (n.Urgent)
        {
            x.Append("<audio silent=\"true\"/>");
        }
        x.Append("</toast>");
        return x.ToString();
    }

    /// <summary>A tag as Windows accepts it: at most 64 UTF-16 units (60 here, to be safe), cut between characters.</summary>
    public static string Tag(string? t)
    {
        if (string.IsNullOrEmpty(t))
        {
            return "";
        }
        var units = 0;
        var sb = new StringBuilder();
        foreach (var r in t.EnumerateRunes())
        {
            if (units + r.Utf16SequenceLength > 60)
            {
                break;
            }
            units += r.Utf16SequenceLength;
            sb.Append(r.ToString());
        }
        return sb.ToString();
    }

    static string Clip(string s, int n)
    {
        var runes = s.EnumerateRunes().ToList();
        return runes.Count <= n ? s : string.Concat(runes.Take(n - 1).Select(r => r.ToString())) + "…";
    }

    /// <summary>
    /// Text made safe as XML text or an attribute value. Characters XML can't carry at all
    /// (most control characters) are dropped: one of them would make Windows refuse the toast.
    /// </summary>
    public static string Esc(string? s)
    {
        var b = new StringBuilder((s ?? "").Length);
        foreach (var r in (s ?? "").EnumerateRunes())
        {
            switch (r.Value)
            {
                case '&':
                    b.Append("&amp;");
                    break;
                case '<':
                    b.Append("&lt;");
                    break;
                case '>':
                    b.Append("&gt;");
                    break;
                case '"':
                    b.Append("&quot;");
                    break;
                case '\'':
                    b.Append("&apos;");
                    break;
                case '\t' or '\n' or '\r':
                    b.Append((char)r.Value);
                    break;
                case < 0x20 or 0xFFFE or 0xFFFF:
                    break; // not allowed in XML 1.0
                default:
                    b.Append(r.ToString());
                    break;
            }
        }
        return b.ToString();
    }
}
